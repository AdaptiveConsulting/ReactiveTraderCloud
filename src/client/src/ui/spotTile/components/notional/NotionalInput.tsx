import numeral from 'numeral'
import React, { PureComponent } from 'react'
import { NUMERAL_FORMAT, isEditMode } from '../TileBusinessLogic'
import {
  InputWrapper,
  CurrencyPairSymbol,
  Input,
  ResetInputValue,
  MessagePlaceholder,
} from './styled'

const DOT = '.'
const ENTER = 'Enter'
const CHAR_CODE_DOT = 46 // .
const CHAR_CODE_0 = 48 // 0
const CHAR_CODE_9 = 57 // 9
const CHAR_CODE_UNIT_SEP = 31 // US (unit separator)
const SHORTCUT_CHAR_CODES = [75, 77, 107, 109] // K, M, k, m

export type ValidationMessage = null | {
  type: 'warning' | 'error' | 'info'
  content: string
}

export interface NotionalUpdate {
  value: string
  type: string
}

interface Props {
  currencyPairSymbol: string
  notional: string
  updateNotional: (notionalUpdate: NotionalUpdate) => void
  resetNotional: () => void
  validationMessage: ValidationMessage
  showResetButton?: boolean
  disabled: boolean
}

interface State {
  showMessage: boolean
}

export default class NotionalInput extends PureComponent<Props, State> {
  private inputRef = React.createRef<HTMLInputElement>()

  handleFocus = (event: React.FocusEvent<HTMLInputElement>) => {
    event.target.select()
  }

  handleKeyPressNotionalInput = (event: React.KeyboardEvent<HTMLInputElement>) => {
    const { charCode, currentTarget, nativeEvent } = event
    if (event.key === ENTER) {
      this.handleChange(event)
    } else if (charCode === CHAR_CODE_DOT) {
      // only allow one dot unless the existing dot is in the current
      // text selection while replacing existing text.
      const { value, selectionStart, selectionEnd } = currentTarget
      const textWithoutSelection = value.replace(value.substring(selectionStart, selectionEnd), '')
      if (textWithoutSelection.match(/\./g)) {
        nativeEvent.stopImmediatePropagation()
        event.preventDefault()
      }
    } else if (!this.inputIsAllowed(charCode)) {
      nativeEvent.stopImmediatePropagation()
      event.preventDefault()
    }
  }

  handleChange = (event: React.FormEvent<HTMLInputElement>) => {
    const { updateNotional } = this.props
    const { currentTarget, type } = event
    const inputValue = currentTarget.value.trim()
    const value = this.formatNotional(inputValue)
    updateNotional({ value, type })
  }

  handleResetNotional = () => {
    this.props.resetNotional()
  }

  formatNotional = (notional: string) => {
    // user may be trying to enter decimals or
    // user may be deleting previous entry (empty string)
    // in those cases, format and update only when completed.
    if (notional === '.') {
      return '0.'
    }
    const lastTwoChars = notional.substr(-2)
    const shouldFormat = !isEditMode(notional) && lastTwoChars.indexOf(DOT) === -1
    return shouldFormat ? numeral(notional).format(NUMERAL_FORMAT) : notional
  }

  inputIsAllowed = (charCode: number) => {
    // allow charcter codes before the Unit Separator to catch Shift, Backspace, etc
    if (charCode <= CHAR_CODE_UNIT_SEP) {
      return true
    }
    // allow shortcut values
    if (SHORTCUT_CHAR_CODES.indexOf(charCode) !== -1) {
      return true
    }
    // allow numeric values:
    if (charCode >= CHAR_CODE_0 && charCode <= CHAR_CODE_9) {
      return true
    }
    return false
  }

  render() {
    const {
      currencyPairSymbol,
      notional,
      validationMessage,
      showResetButton,
      disabled,
    } = this.props

    return (
      <InputWrapper altLayout={Boolean(validationMessage)}>
        <CurrencyPairSymbol>{currencyPairSymbol}</CurrencyPairSymbol>
        <Input
          type="text"
          ref={this.inputRef}
          value={notional}
          onFocus={this.handleFocus}
          onChange={this.handleChange}
          onBlur={this.handleChange}
          onKeyPress={this.handleKeyPressNotionalInput}
          validationMessage={validationMessage}
          disabled={disabled}
        />
        {showResetButton && (
          <ResetInputValue onClick={this.handleResetNotional}>
            <i className="fas fa-redo fa-flip-horizontal" />
          </ResetInputValue>
        )}
        {validationMessage && (
          <MessagePlaceholder validationMessageType={validationMessage.type}>
            {validationMessage.content}
          </MessagePlaceholder>
        )}
      </InputWrapper>
    )
  }
}
