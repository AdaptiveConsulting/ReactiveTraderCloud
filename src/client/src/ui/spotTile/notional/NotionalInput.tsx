import * as React from 'react'
import * as numeral from 'numeral'
import { utils } from '../../../system'
import * as classnames from 'classnames'
import './notionalInput.scss'

const NUMERAL_FORMAT      = '0,000,000[.]00'
const DOT                 = '.'
const ENTER               = 'Enter'
const CHAR_CODE_DOT       = 46
const CHAR_CODE_0         = 48
const CHAR_CODE_9         = 57
const CHAR_CODE_UNIT_SEP  = 31
const SHORTCUT_CHAR_CODES = [75, 77, 107, 109] // K, M, k, m

const MAX_NOTIONAL_VALUE = 1000000000

export interface CurrencyPair {
  symbol: string
  base: string
}

export interface NotionalInputProps {
  className: string
  notional: number
  currencyPair: CurrencyPair
  onNotionalInputChange: (value: number) => void
}

export default class NotionalInput extends React.Component<NotionalInputProps, {}> {
  props: NotionalInputProps
  refs: any
  shouldComponentUpdate(nextProps: any, nextState: any) {
    return this.props.className !== nextProps.className ||
      this.props.notional !== nextProps.notional ||
      // currencyPair always here
      this.props.currencyPair.symbol !== nextProps.currencyPair.symbol ||
      this.props.onNotionalInputChange !== nextProps.onChange
  }

  render() {
    const { className, currencyPair, notional } = this.props

    const formattedSize = numeral(notional).format(NUMERAL_FORMAT)
    const classes = classnames('notional', className)

    return (
      <div className={classes}>
        <label className="notional__currency-pair">{currencyPair.base}</label>
        <input className="notional__size-input"
          type="text"
          ref="notionalInput"
          defaultValue={formattedSize}
          onClick={this.handleSelect}
          onChange={(e: any) => this.handleInputChange(e)}
          onBlur={(e: any) => this.processNotional(e.target.value)}
          onKeyPress={(e: any) => this.handleKeyPressNotionalInput(e)}/>
      </div>
    )
  }

  handleKeyPressNotionalInput(e: any) {
    const charCode = e.charCode

    if (e.key === ENTER) {
      this.processNotional(e.target.value)
    } else if (charCode === CHAR_CODE_DOT) {
      // only allow one dot
      const numDots = e.target.value.match(/\./g).length
      if (numDots > 0) {
        e.nativeEvent.stopImmediatePropagation()
        e.preventDefault()
      }
    } else if (!this.inputIsAllowed(charCode)) {
      e.nativeEvent.stopImmediatePropagation()
      e.preventDefault()
    }
  }

  processNotional(inputValue: string) {
    const inputValueTrimmed = inputValue.trim()
    let notional: any = utils.convertNotionalShorthandToNumericValue(inputValueTrimmed)
    if (notional >= MAX_NOTIONAL_VALUE) {
      notional = 0
    }
    if (!isNaN(notional)) {
      // send temp notional back to parent
      this.props.onNotionalInputChange(notional)

      // user may be trying to enter decimals. restore BACK into input
      if (inputValueTrimmed.indexOf(DOT) === inputValueTrimmed.length - 1) {
        notional = notional + DOT
      }
      // propagate change back to dom node's value
      this.refs.notionalInput.value = numeral(notional).format(NUMERAL_FORMAT)
    }
  }

  handleSelect(e: any) {
    const el = e.target
    el.setSelectionRange(0, el.value.length)
  }

  handleInputChange(e: any) {
    const rawValue = (this.refs.notionalInput.value || e.target.value).trim()
    // check for a shortcut input
    if (utils.hasShorthandInput(rawValue)) {
      this.processNotional(rawValue)
    }
  }

  inputIsAllowed(charCode: number) {
    // allow charcter codes before the Unit Separator to catch Shift, Backspace, etc
    if (charCode <= CHAR_CODE_UNIT_SEP) return true

    // allow shortcut values
    if (SHORTCUT_CHAR_CODES.indexOf(charCode) !== -1) return true

    // allow numeric values:
    if (charCode >= CHAR_CODE_0 && charCode <= CHAR_CODE_9) return true
    return false
  }
}
