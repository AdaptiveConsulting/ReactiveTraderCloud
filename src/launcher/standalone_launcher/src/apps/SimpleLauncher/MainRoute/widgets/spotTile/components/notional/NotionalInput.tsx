import React, { PureComponent } from 'react'
import {
  InputWrapper,
  CurrencyPairSymbol,
  Input,
  ResetInputValue,
  MessagePlaceholder,
} from './styled'
import { ValidationMessage } from './types'

interface Props {
  currencyPairBase: string
  currencyPairSymbol: string
  notional: number
  updateNotional: (notional: number) => void
  resetNotional: () => void
  validationMessage?: ValidationMessage
  showResetButton?: boolean
  disabled: boolean
}

interface State {
  showMessage: boolean
}

export default class NotionalInput extends PureComponent<Props, State> {
  private inputRef = React.createRef<HTMLInputElement>()

  handleFocus = (event: React.FocusEvent<HTMLInputElement>) => {
    event.target.select()
  }

  handleResetNotional = () => {
    this.props.resetNotional()
  }

  render() {
    const {
      currencyPairSymbol,
      currencyPairBase,
      notional,
      updateNotional,
      validationMessage,
      showResetButton,
      disabled,
    } = this.props

    return (
      <InputWrapper altLayout={Boolean(validationMessage)}>
        <CurrencyPairSymbol data-qa="notional-input__currency-pair-symbol">
          {currencyPairBase}
        </CurrencyPairSymbol>
        <Input
          type="text"
          ref={this.inputRef}
          numericValue={notional}
          onFocus={this.handleFocus}
          onNumericValueChange={updateNotional}
          validationMessage={validationMessage}
          disabled={disabled}
          data-qa-id={`notional-input__input-${currencyPairSymbol.toLowerCase()}`}
        />
        {showResetButton && (
          <ResetInputValue
            onClick={this.handleResetNotional}
            data-qa="notional-input__reset-input-value"
          >
            <i className="fas fa-redo fa-flip-horizontal" />
          </ResetInputValue>
        )}
        {validationMessage && (
          <MessagePlaceholder
            validationMessageType={validationMessage.type}
            data-qa="notional-input__validation-message"
          >
            {validationMessage.content}
          </MessagePlaceholder>
        )}
      </InputWrapper>
    )
  }
}
