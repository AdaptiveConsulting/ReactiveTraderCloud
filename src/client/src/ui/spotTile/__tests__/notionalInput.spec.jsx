/* tslint:disable */

import React from 'react'
import NotionalInput from '../notional/NotionalInput'
import ShallowRenderer from 'react-test-renderer/shallow'
import { mount } from 'enzyme'


describe.skip('NotionalInput', () => {
  let component
  let mockAction

  beforeEach(() => {
    mockAction = jest.fn(() => {})
    component = <NotionalInput
      className={'notional'}
      notional={'999999K'}
      currencyPair={{
        symbol: 'GBP',
        base: 'GBP',
      }}
      onNotionalInputChange={mockAction}
    />
  })

  test('renders correct markup', () => {
    const shallowRenderer = new ShallowRenderer()
    shallowRenderer.render(component)
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot()
  })

  test('onNotionalInputChange triggers when using shorthand value with "k"', () => {
    const mounted = mount(component)
    mounted.find('input').simulate('change', '12k')
    expect(mockAction).toBeCalled()
  })

  test('onNotionalInputChange triggers when using shorthand value with "m"', () => {
    const mounted = mount(component)
    mounted.find('input').simulate('change', '10m')
    expect(mockAction).toBeCalled()
  })

  test('onNotionalInputChange does not trigger for regular values', () => {
    const mounted = mount(component)
    mounted.find('input').get(0).value = '10'
    mounted.find('input').simulate('change')
    expect(mockAction).not.toBeCalled()
  })
})
