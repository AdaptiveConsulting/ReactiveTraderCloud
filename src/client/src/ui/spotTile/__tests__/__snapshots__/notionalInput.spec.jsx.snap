// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotionalInput renders correct markup 1`] = `
<div
  className="notional notional"
>
  <label
    className="notional__currency-pair"
  >
    GBP
  </label>
  <input
    className="notional__size-input"
    defaultValue="999,999"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onKeyPress={[Function]}
    type="text"
  />
</div>
`;
